overleaf/real-time
# [backport] 113 and 126: ol_context

### Description

This PR backports the synchronous client context `ol_context` before we cleanup whitespace in the decaff process #101 . #113 adds the core changes and #126 makes the cleanup more robust.

(We saw events coming in after clients had passed though the disconnect process, this was likely the result of `process.nextTick` usage in v2 and does not affect v0. Making v0 more robust should not hurt, so I chose to include these patches in this PR.)

#### Related Issues / PRs

#113 
#126

### Review

Three tests cases/suites related to the early disconnect #145 and update size limitation #120 required additional changes :

<details>

- `WebsocketController` -> `leaveProject` -> `when the client did not joined a project yet`

   reset `client.ol_context` instead of `client.params`

- `WebsocketController` -> `leaveProject` -> `should not inc any metric`

   reset `client.ol_context` instead of `client.params`

- `WebsocketController` -> `update_too_large` -> `before each`

   set `user_id`/`project_id` on `client.ol_context` instead of `client.params`

</details>

#### Potential Impact

High. Restricted client could get all messages. We have acceptance tests for that. All the other context access is covered by acceptance tests and unit tests by now.

#### Manual Testing Performed

- run unit/acceptance tests
- open project in dev-env
- verify that token-access users can not see chat messages (which is also covered by tests)
